### ISSUE TITLE
Accuracy Drop with Flash-Attention Reimplementation in Encoder-Decoder Architecture (ViT)

### ISSUE THREAD

**Issue by ImaGonEs** [Dao-AILab/flash-attention#1376]
Recently, I did a flash-attention reimplementation of a Visual Encoder-Decoder architecture similar to the one presented in [MAGE](https://github.com/LTH14/mage/blob/main/models_mage.py). My implementation introduces some additional losses but otherwise retains the general structure. 

## Implementation 
In an attempt to increase the efficiency, I replaced the Attention and Block classes with the following adaptations: 

```python
class FlashAttentionLayer(nn.Module):
    def __init__(self, dim, num_heads=8, qkv_bias=False, qk_scale=None, attn_drop=0., proj_drop=0.):
        super().__init__()
        self.num_heads = num_heads
        head_dim = dim // num_heads
        self.scale = qk_scale or head_dim ** -0.5
        self.attn_drop_prob = attn_drop
        
        self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
        self.proj = nn.Linear(dim, dim)
        self.proj_drop = nn.Dropout(proj_drop)

    def forward(self, x, causal=False):
        B, N, C = x.shape
        qkv = self.qkv(x).reshape(B, N, 3, self.num_heads, C // self.num_heads)

        
        attn_output = flash_attn_qkvpacked_func(
                qkv, 
                dropout_p=self.attn_drop_prob if self.training else 0.0, 
                softmax_scale=self.scale, 
                causal=causal
            )        

       
        x = attn_output.permute(0, 2, 1, 3).reshape(B, N, C)
        x = self.proj(x)
        x = self.proj_drop(x)
        return x
```

```python
class FlashBlock(nn.Module):
    def __init__(self, dim, num_heads, mlp_ratio=4., qkv_bias=False, qk_scale=None, drop=0., attn_drop=0.,
                drop_path=0., act_layer=nn.GELU, norm_layer=nn.LayerNorm):
        super().__init__()
        self.norm1 = norm_layer(dim)
        self.attn = FlashAttentionLayer(
            dim, num_heads=num_heads, qkv_bias=qkv_bias, qk_scale=qk_scale, attn_drop=attn_drop, proj_drop=drop
        )
        self.drop_path = DropPath(drop_path) if drop_path > 0. else nn.Identity()
        self.norm2 = norm_layer(dim)
        mlp_hidden_dim = int(dim * mlp_ratio)
        self.mlp = Mlp(in_features=dim, hidden_features=mlp_hidden_dim, act_layer=act_layer, drop=drop)

    def forward(self, x, return_attention=False):
        if return_attention:
            raise ValueError("We do not return attention maps.")
        else:
            y = self.attn(self.norm1(x))
            x = x + self.drop_path(y)
            x = x + self.drop_path(self.mlp(self.norm2(x)))
        return x
```

## Results 

Even if I got nice speedup, when running and comparing both (even if is SSL training, I have an online classifier implemented), the accuracy drops to 10 points just by changing normal Blocks with FlashBlocks. I use the same scheduler, optimizer (AdamW) and hyperparameters (dropout included) for both approaches:

```
    --mask_ratio_min 0.5 --mask_ratio_max 1.0 \
    --mask_ratio_mu 0.55 --mask_ratio_std 0.25 \
    --epochs 200 \
    --warmup_epochs 10 \
    --blr 1.5e-4 --weight_decay 0.05 \
```

These are the plots for both implementations (the sudden spike is due to an additional loss):

![image](https://github.com/user-attachments/assets/1fd6e8da-98a9-4f21-acb7-6abee53d8dc6)

## Question

Is there any known instability when using Flash-Attention in these Encoder-Decoder architectures? Specifically:

1. Does Flash-Attention require additional tuning (e.g., different learning rates, dropout rates, or initialization schemes)?
2. Could this be an issue related to number stability or implementation-specific nuances? I see 10 points difference as a huge gap.

Thanks in advance!

**Comment by tridao** [comment]
I don't think this is right
```
x = attn_output.permute(0, 2, 1, 3).reshape(B, N, C)
```

**Comment by ImaGonEs** [comment]
Solved! I'm still getting slightly different results but both approaches follow the same line now. 

Thanks!!
